### PR TITLE
Convert TF models using TF Python API.

### PR DESCRIPTION
**Pull request by dellis23**
Not ready for review yet. I need to test the tool itself with the benchmarks.

### CODE REVIEW

**Comment by dellis23** [comment]
@mariecwhite I'd like to work on getting the benchmarks to use the new tool / making sure it works correctly.  Any code pointers you can give on how the benchmarks are setup and how I can get them to run on my PR?  I'll be creating a parallel tool for now so we can test with customers before deleting the old tool, so I'll need to modify the way the benchmarks are running the tool.

**Comment by mariecwhite** [comment]
For info on how the tool is used in the IREE repo, @pzread would know more.

For SHARK, it pip installs `iree-import-tf` from the nightly releases. Here are instructions on how to use SHARK: https://gist.github.com/mariecwhite/540279108151f4e4e249ca9fb307d6f0#shark. There is a way to install locally built versions of IREE tools and they're detailed in the instructions. When running the pytests, I suggest using `cpu and tf and static` for `BENCHMARK_QUERY` for local testing.

**Comment by pzread** [comment]
> @mariecwhite I'd like to work on getting the benchmarks to use the new tool / making sure it works correctly. Any code pointers you can give on how the benchmarks are setup and how I can get them to run on my PR? I'll be creating a parallel tool for now so we can test with customers before deleting the old tool, so I'll need to modify the way the benchmarks are running the tool.

If this simply replaces the tool `iree-import-tf` built from `integrations/tensorflow`, then adding `benchmarks:*` labels should be able to test it with the IREE benchmark suites (for each commit, we build the `iree-import-tf` from `integrations/tensorflow` and use it to import models for the benchmarks)

File: compiler/bindings/python/iree/compiler/tools/tf.py
Comment (stellaraccident) [inline compiler/bindings/python/iree/compiler/tools/tf.py:218]: that's weird. I probably would have just used the contents directly vs redirecting through `cat`

File: integrations/tensorflow/iree_tf_compiler/BUILD.bazel
Comment (GMNGeoffrey) [inline integrations/tensorflow/iree_tf_compiler/BUILD.bazel]: Just delete this. There is no need to "build" a binary for Python. We have elected to completely bail on trying to use Bazel for Python dependency management. It may be desirable to move the tool into a separate directory so that it's clearly distinct from these compile C++ binaries (which should also go away soon). Then there are no symlinks: the CMake build just directly references the Python executable script.

Comment (dellis23) [reply integrations/tensorflow/iree_tf_compiler/BUILD.bazel]: (I assume this is a slightly delayed comment, so resolving. I removed this in the previous commit.)

File: .github/workflows/ci.yml
Comment (ScottTodd) [inline .github/workflows/ci.yml:677]: This is rough around the edges, but I was just able to reproduce what the CI does locally with this sequence of commands: https://gist.github.com/ScottTodd/bcb137207c7320865442f024e61c61d2

The trickiest details are in how Docker handles mounted directories and how artifacts are copied between jobs

Comment (dellis23) [reply .github/workflows/ci.yml:677]: Re: TF_BINARIES_OUTPUT_DIR, what commit or branch did you run this on?  The new tool doesn't end up in TF_BINARIES_OUTPUT_DIR, since we're no longer building with bazel / copying around symlinks.  It's installed when the package is pip installed (although, as you can see from my commits, the environment still isn't aware of the installed commands).

Comment (ScottTodd) [reply .github/workflows/ci.yml:677]: I'm forked off of https://github.com/openxla/iree/pull/12758/commits/226434815b15bce74e0b4d58d518bca2e3827fe1.

Working here: https://github.com/ScottTodd/iree/tree/tf-python-api

For local iteration I've commented out most of `tests/e2e/test_artifacts/generated_e2e_test_iree_artifacts.cmake`, so I can just run the `build_e2e_test_artifacts.sh` step and see a single set of error messages from `iree_import_tf_model` (next I was going to try the pip install commands you added)

Comment (dellis23) [reply .github/workflows/ci.yml:677]: I need to disable 5 tests in `generated_e2e_test_iree_artifacts.cmake`, but I assume it's not a good idea to modify that file directly.  I've tried to traverse the numerous layers involved but haven't figured out the core source of truth yet.  Going to continue on it tomorrow.

Comment (ScottTodd) [reply .github/workflows/ci.yml:677]: instructions: https://github.com/openxla/iree/tree/main/tests/e2e/test_artifacts#updating-test-artifacts (and maybe @pzread or someone else could help you with the specific targets)

Comment (dellis23) [reply .github/workflows/ci.yml:677]: Is something going on with github CI right now? I can't get anything but the CLA agreement test to run.  Only 7 other jobs are showing up, and they all say `Expected — Waiting for status to be reported`.

**Comment by dellis23** [comment]
I believe this is ready for review.  Additional context for those that haven't been in all of the discussions: @jpienaar and @stellaraccident have expressed a desire to sever this dependency; for now, that means removing some failing tests and removing the failing benchmarks.  It should be possible to fix the problems and re-enable the benchmarks, it's just unclear how much work will be involved.

I haven't yet removed all of the actual build dependencies.  I will do that in a followup PR.  I may need some pointers to rip all of that out and make sure we're actually getting the lower build times we're hoping for.

File: integrations/tensorflow/test/requirements.txt
Comment (jpienaar) [inline integrations/tensorflow/test/requirements.txt]: Should we pin to a specific one?

Comment (dellis23) [reply integrations/tensorflow/test/requirements.txt]: I guess we could temporarily.  All we're doing really is waiting for the next TF release, at which point we should be able to get back on the main TF releases.

Comment (jpienaar) [reply integrations/tensorflow/test/requirements.txt]: Yes, my concern is just that the testing of nightly is lower, so I'd prefer not flakes from TF updates and I'm ok to pinning to any working one until next release.

**Review by jpienaar — APPROVED** [review]

File: integrations/tensorflow/test/requirements.txt
Comment (dellis23) [reply integrations/tensorflow/test/requirements.txt]: Makes sense, updated

**Review by stellaraccident — APPROVED** [review]
I deem the drop in test coverage a necessary tradeoff for moving forward. In practice, the TFLite and Jax test suites have long since been the high ground here driving current work. I will note that, in theory, most of this test coverage could be retained with a re-engineering of the TF regression test tooling to make it not assume a 1:1 correspondence between the TF-presented API and the IREE-exposed API. This would make it fairly similar to the level of support that tf-compile has (as an example), but with state/variables modeled.

The will to keep this programming model alive has long since lapsed, and we are better spending our time on current generation testing efforts and framework features.

File: compiler/bindings/python/iree/compiler/tools/tf.py
Comment (dellis23) [reply compiler/bindings/python/iree/compiler/tools/tf.py:218]: Oh weird, somehow I missed this comment.  Yeah, I considered reworking that too but was just wanting to get something working and was trying to re-used the existing command execution pipeline code.  Let me know if you want me to read and feed through stdin in Python instead.

Comment (stellaraccident) [reply compiler/bindings/python/iree/compiler/tools/tf.py:218]: Can be fixed post. I expect this will fail on Windows, but none of this has worked on Windows in a long time, afaik.

**Comment by mariecwhite** [comment]
Great work on getting this merged. However it removed a bunch of models from the benchmarks. Can you re-add them?

**Comment by antiagainst** [comment]
This breaks the Android benchmark pipeline: https://buildkite.com/iree/iree-benchmark-android/builds/6478. https://github.com/openxla/iree/pull/12964 to fix.

File: .github/workflows/ci.yml
Comment (GMNGeoffrey) [reply .github/workflows/ci.yml:677]: Should this step have been removed before merging?

**Review by GMNGeoffrey — COMMENTED** [review]
Yeah I'm concerned that this dropped a whole bunch of our benchmarks, many of which were just added recently (not to mention tests). If the plan is to fix forward, who owns doing that? Who owns chasing down the performance regressions we will miss until that completes?

**Comment by julianwa** [comment]
> Great work on getting this merged. However it removed a bunch of models from the benchmarks. Can you re-add them?

+1. What's the plan here?

**Comment by mariecwhite** [comment]
I tried adding the TF models back and now getting new failures. Reverting this PR.

**Comment by dellis23** [comment]
Hey everyone,

I'll defer to @stellaraccident and @jpienaar on the decision re: the tradeoffs and plan going forward.  In case some of the comments got lost in the noise, Stella put a bit more context [here](https://github.com/openxla/iree/pull/12758#pullrequestreview-1363607184), and I tried to type up some of the background [here](https://github.com/openxla/iree/pull/12758#issuecomment-1499392151) as well.

Thanks @GMNGeoffrey for catching that debug line. I've removed in https://github.com/openxla/iree/pull/12973.

**Comment by stellaraccident** [comment]
> > Great work on getting this merged. However it removed a bunch of models from the benchmarks. Can you re-add them?
> 
> +1. What's the plan here?

The plan is to fix forward for things that are important and drop the rest. We cannot rely on an API to tensorflow that is wider than what is upstream and folks are going to have to adapt. I appreciate that this will cause some rework.

**Comment by julianwa** [comment]
> The plan is to fix forward for things that are important and drop the rest. We cannot rely on an API to tensorflow that is wider than what is upstream and folks are going to have to adapt. I appreciate that this will cause some rework.

We're actively adding these models to the benchmarking suite to do nightly runs to drive our dashboards and monthly benchmark reviews. It's one of @mariecwhite's top priorities. So not only does this reverse some of that work, it blocks her from making progress. I don't think it's responsible to land this without that solution in place. I think we're fine adapting. But I want to know what the timeline is, what the work is that remains, who's going to do it, and why we shouldn't land it before submitting this.

**Comment by stellaraccident** [comment]
There are two things: a) who is doing the work to pick what carries forward and getting it working again, and b) are people truly blocked or has the cheese been moved.

For (a) I would assume that Dan will pick through the remains and work on them.
For (b) I'm willing to be proven wrong, but I think the API got uglier (for some definition of ugly) but it is still usable. This is par for the course with TensorFlow model extraction APIs and will require adaptation.

**Comment by julianwa** [comment]
> For (b) I'm willing to be proven wrong, but I think the API got uglier (for some definition of ugly) but it is still usable. This is par for the course with TensorFlow model extraction APIs and will require adaptation.

I think @mariecwhite is blocked in the sense that she tried to restore the benchmarks and it didn't work using her initial approach. I assume this should be blocked until we hear what it will take to restore them (i.e., where is the new cheese). If we have a better understanding, we can decide whether to fix first or fix forward. I'm assuming @dellis23 will help us figure this out? @mariecwhite's weekend has already started.

**Comment by stellaraccident** [comment]
In any case, the PR is reverted and Jacques will be online after some commitments.

To be honest, I may have been level setting on my assumption that "TensorFlow interfacing is terrible, unpredictable, and I can always fudge it into working in a few minutes" as the baseline in terms of what is a simple thing to do or not. I realize that not everyone has mind-melded with the beast to the same extent...

I think a way forward here is likely for the author (Dan) to retrofit one of the benchmarks and then pose the question on how to proceed.

For the unit tests, we're not going to fix those. Not worth the cost-benefit at this point in time given the other testing sources and legacy of those tests.

**Comment by stellaraccident** [comment]
(I'll also note that maintaining this kind of interface for TF is not on anyone's priorities and we would be best invested in supported mechanisms that are convergent with our roadmap vs investing heavily in TF-based testing infra. But I realize we are in the "the future isn't ready yet" state, so this is just a caution at this point)

**Comment by jpienaar** [comment]
To give some context here. The TF import API used by the benchmarks have been deprecated for about a year at this point. They are not maintained nor anyone's responsibility (any fixes you have seen have been done by folks going out of their way). Dan spent roughly a month here migrating us to the new API's that TF org have suggested we move to. These APIs have less functionality (e.g., you may have to freeze your graph or you may have to add special annotation in Python before invoking) than the previous, but those are the APIs suggested. He was doing tech debt reduction here following the roadmap and the prioritization agreed upon.

I'm not sure why TF benchmarks are currently Marie's main priority. TF benchmark being functional and using PJRT as supported API is by the end of 2023 on the roadmap. For the ML framework specific integration (independent of framework) we plan to use JIT/PJRT. The current focus is JAX and PyTorch, work to add benchmarks from these frameworks or enabling adding them is priority. Combining with the previous, if we benchmark, it should be via PJRT or benchmarking the StableHLO module (serialized/dumped using MLIR bytecode) - the latter has the advantages that it can be done today already for JAX, PyTorch and TF in a uniform way, it would be much cheaper, lower dependency cost and isolate compiler & runtime regressions/improvements from framework ones. Doesn't mean we don't need full integration tests, but currently frameworks are pinned per release so we don't actually get any signal of framework related issues here (iree-samples and iree-jax does that in a limited way, but nothing for TF, and I don't think we want a framework dep on openxla/iree).

(Aside: We are working on adding more testing and verification openxla-prjrt-plugin side, we have 4 issues open that we want to address first before we do. Benchmarks could be added there post.)

The TF specific code here is in contrast with how PyTorch and JAX is handled and not aligned with where we are going (e.g., PJRT plugin is in its own repo, not here exactly because we don't want all the things of all the frameworks here). To approximate closest, our state wrt TensorFlow is a StableHLO compiler and we test StableHLO compilation and performance. We have this fully AOT flow today and we need to rebase this on supported APIs. For benchmarking & flagging regressions for TF models the serialized StableHLO modules are best option and closest aligned to future state.

**Comment by GMNGeoffrey** [comment]
I'm very much in favor of severing our TF dependency. I think the issue here was one of communication: people were caught off guard. We should definitely migrate things we care about to a supported API. `iree-import-tf` is how we've imported TF models since forever though and it's never been a supported API because TF didn't have one and we just hacked something together. I knew that Dan was working on getting us off that (and was eagerly anticipating the results), but I was not aware of any actual deprecation of some import path or request to make the benchmarks (or anything else) compatible with a different API, and the first I heard about them breaking was when Lei sent me a PR to fix the android pipeline and I happened to look at the PR that caused it to break. Based on the reactions, I think I'm not alone in that.

It sounds like there's something relatively straightforward (well as straightforward as anything in TF) we can do to make the models we're importing compatible with this more limited API. Can we please have, say, a week to migrate these *before* this relands? @mariecwhite, when you're back on Tuesday, does that give you enough time to freeze the graphs for these input models?

**Comment by mariecwhite** [comment]
We need TF benchmarks to make comparisons between IREE and XLA, which is a priority. I would be more than happy to use a mechanism that doesn't include the TF framework but we need to be running the existing benchmarks until a suitable replacement is ready, otherwise there will be gaps in our longitudinal data.

**Comment by ScottTodd** [comment]
My 2c: the TF import code and everything we have depending on it is a big Gordian Knot. Any way that we try to unravel it or cut it is going to be difficult, but the benefits outweigh the costs. We had a few discussions about this work on this PR and on Discord, but some signal boost would have helped before merging with known issues.

Maybe for migrating, we could create `iree-import-tf-2` using the Python API, keeping the old code and binary around for a bit? Then we can iterate on the infrastructure updates on the `main` branch?

**Comment by ScottTodd** [comment]
Oh, and while the commit had some description (https://github.com/openxla/iree/commit/164bad391bee2f94aab05ea1f8dc5ef11a88408e), that could have been expanded upon to include known issues and other context. That information should also be in the PR description (which is currently just "Not ready for review yet. I need to test the tool itself with the benchmarks.").

**Comment by dellis23** [comment]
I'm skeptical that updating the original comment would have fixed any of these issues.  There are a number of places in the code here where I've tried to point to the rationale for this change.  I've linked people to the multiple comments explaining this too.  I've tried to make it clear that removing the benchmarks was not an incidental side effect of this, it was deemed a necessary tradeoff and is indeed the point of the PR.  Despite all this, I've gotten repeated private messages that I "broke the benchmarks."  

At every step of the way I've had people pinging me to get this done as soon as possible because of how important it was deemed to be (not too much pinging, to be clear; a reasonable amount that communicated importance and urgency).  I completely understand every viewpoint expressed here, and I don't disagree in principle with any of them.  It's ultimately a question of tradeoffs.

I don't really like being in the middle of any of this, as I'm being pulled in a number of different directions and working on something that frankly hasn't been pleasant to work on even if none of this had happened.  Despite that, I was willing to fix every last one of these tests and benchmarks before submitting if that was deemed necessary (but it wasn't!).  I had no skin in the game.

There is a clear elephant in the room here about the organization of our priorities, our roadmap, and our decision making.  As I see it, every proposed technical solution here is secondary to that.

**Comment by ScottTodd** [comment]
Agreed! I'd have liked to see those affected engage with this work earlier and more actively. It's very difficult but necessary work and we should be finding ways to share that load. There are definitely some process / communication system glitches here to work through.

**Comment by stellaraccident** [comment]
I was caught unaware that we were coupling new things to this path (in my mind, we announced this was going away a year ago). We probably have some steering to do on the benchmark side, but I think we can square this issue... I have a couple of hours today that I can get caught up. I am motivated to see this issue resolved.

**Comment by stellaraccident** [comment]
(also - plug: CI and regular builds are now running for the PJRT plugin/Jax, and that is where a lot of this stuff needs to migrate)

**Comment by mariecwhite** [comment]
We are generating TF saved models here: https://github.com/iree-org/iree-samples/blob/main/iree-tf/library/generate_saved_models.py#L53. If you can make the necessary changes to the SavedModel generation to make it work with the new API, I can do the rest (updating the benchmark artifacts, etc.).

**Pull request merged**
